{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"machine_shape":"hm","authorship_tag":"ABX9TyOVAIUbmyjIPK4eqM+bGKPN"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"7k54usNVV-AO","executionInfo":{"status":"ok","timestamp":1665819611968,"user_tz":-540,"elapsed":6928,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"96c07be2-c6f9-44f4-9fe9-7f2d705320e3"},"outputs":[{"output_type":"stream","name":"stdout","text":["\u001b[K     |████████████████████████████████| 15.1 MB 3.8 MB/s \n","\u001b[?25h"]}],"source":["!pip install -q pysam"]},{"cell_type":"code","source":["import pysam\n","pysam.__version__"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":36},"id":"JByah8o0WgNM","executionInfo":{"status":"ok","timestamp":1665819611969,"user_tz":-540,"elapsed":16,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"f96b75f1-4447-4d2c-fc5c-de5d87b73fc8"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'0.19.1'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":2}]},{"cell_type":"markdown","source":["- https://hgdownload.soe.ucsc.edu/downloads.html\n","\n","- http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/\n","\n","- wgEncodeUwRepliSeqBg02esG1bAlnRep1.bam\tproject=wgEncode; grant=Stam; lab=UW; composite=wgEncodeUwRepliSeq; dataType=RepliSeq; view=Alignments; cell=BG02ES; phase=G1b; replicate=1; origAssembly=hg19; dataVersion=ENCODE Mar 2012 Freeze; dccAccession=wgEncodeEH002250; dateSubmitted=2011-11-01; subId=5040; labExpId=DS9027; type=bam; md5sum=44c2a0c9df43fee5638e408138d9fce9; size=92M\n","\n","- BG02ES is a cell line dervied from human embryonic stem cells.\n","- https://www.ebi.ac.uk/ols/ontologies/efo/terms?short_form=EFO_0002778"],"metadata":{"id":"2SKuqmlEuKPN"}},{"cell_type":"code","source":["import urllib\n","from urllib import request\n","\n","src_file = \"http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwRepliSeq/wgEncodeUwRepliSeqBg02esG1bAlnRep1.bam\"\n","dest_file = 'BG02ES.bam'\n","\n","urllib.request.urlretrieve(src_file, dest_file)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"9pHqNxNxYmnJ","executionInfo":{"status":"ok","timestamp":1665819619088,"user_tz":-540,"elapsed":7131,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"7b9e3b86-dcfb-478b-b19b-ba441525809c"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["('BG02ES.bam', <http.client.HTTPMessage at 0x7fecb40f6750>)"]},"metadata":{},"execution_count":3}]},{"cell_type":"code","source":["import pysam\n","file_path = \"/content/BG02ES.bam\"\n","samfile = pysam.AlignmentFile(file_path, \"rb\")"],"metadata":{"id":"e5gpslGSWnn8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(samfile.header)\n","\n","# SQ Reference sequence dictionary\n","# SN Reference sequence name \n","# LN Reference sequence length"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TVosMyXmyQ2Y","executionInfo":{"status":"ok","timestamp":1665819619089,"user_tz":-540,"elapsed":10,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"000c3b4a-9078-4a3e-cee1-007a31ad3757"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","@SQ\tSN:chr1\tLN:249250621\n","@SQ\tSN:chr2\tLN:243199373\n","@SQ\tSN:chr3\tLN:198022430\n","@SQ\tSN:chr4\tLN:191154276\n","@SQ\tSN:chr5\tLN:180915260\n","@SQ\tSN:chr6\tLN:171115067\n","@SQ\tSN:chr7\tLN:159138663\n","@SQ\tSN:chr8\tLN:146364022\n","@SQ\tSN:chr9\tLN:141213431\n","@SQ\tSN:chr10\tLN:135534747\n","@SQ\tSN:chr11\tLN:135006516\n","@SQ\tSN:chr12\tLN:133851895\n","@SQ\tSN:chr13\tLN:115169878\n","@SQ\tSN:chr14\tLN:107349540\n","@SQ\tSN:chr15\tLN:102531392\n","@SQ\tSN:chr16\tLN:90354753\n","@SQ\tSN:chr17\tLN:81195210\n","@SQ\tSN:chr18\tLN:78077248\n","@SQ\tSN:chr19\tLN:59128983\n","@SQ\tSN:chr20\tLN:63025520\n","@SQ\tSN:chr21\tLN:48129895\n","@SQ\tSN:chr22\tLN:51304566\n","@SQ\tSN:chrM\tLN:16571\n","@SQ\tSN:chrX\tLN:155270560\n","\n"]}]},{"cell_type":"code","source":["print(samfile.references)\n","print(samfile.nreferences)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aWVdurxXyWYE","executionInfo":{"status":"ok","timestamp":1665819619089,"user_tz":-540,"elapsed":8,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"7617c95f-d812-45dc-c711-7c38319d7178"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["('chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrM', 'chrX')\n","24\n"]}]},{"cell_type":"code","source":["print(samfile.lengths)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"A2h0zubBcgb1","executionInfo":{"status":"ok","timestamp":1665819619090,"user_tz":-540,"elapsed":8,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"50b29365-3ede-47bd-8300-4cb16210e3ab"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663, 146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540, 102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566, 16571, 155270560)\n"]}]},{"cell_type":"code","source":["# idex作成\n","file_path = \"/content/BG02ES.bam\"\n","pysam.index(file_path)\n","samfile = pysam.AlignmentFile(file_path, \"rb\")"],"metadata":{"id":"6CLHrSH_666B"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["samfile.count(\"chr1\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yDYaZ-TAzWAE","executionInfo":{"status":"ok","timestamp":1665819621330,"user_tz":-540,"elapsed":1116,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"9a1cfb14-641e-4fef-d12b-4f0faa7229d7"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["179687"]},"metadata":{},"execution_count":9}]},{"cell_type":"code","source":["samfile.fetch( )\n","\n","# (contig: str | None = ..., start: int | None = ..., stop: int | None = ..., region: str | None = ..., tid: int | None = ..., until_eof: bool = ..., multiple_iterators: bool = ..., reference: str | None = ..., end: int = ...) -> IteratorRow\n","# AlignmentFile.fetch(self, contig=None, start=None, stop=None, region=None, tid=None, until_eof=False, multiple_iterators=False, reference=None, end=None)\n","# fetch reads aligned in a :term:`region`."],"metadata":{"id":"yjVda9D_4Xse","executionInfo":{"status":"ok","timestamp":1665819621331,"user_tz":-540,"elapsed":11,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"52e5e6b5-ad26-41c8-be4b-a1c5d24afd45"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<pysam.libcalignmentfile.IteratorRowAllRefs at 0x7fecb411d050>"]},"metadata":{},"execution_count":10}]},{"cell_type":"code","source":["samfile = pysam.AlignmentFile(file_path, \"rb\")\n","for idx, read in enumerate(samfile.fetch(reference='chr1')):\n","  if idx < 5:\n","    print(read)\n","\n","samfile.close()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DX9X03VhcveP","executionInfo":{"status":"ok","timestamp":1665819621331,"user_tz":-540,"elapsed":9,"user":{"displayName":"Jing Wora","userId":"14098338404840441752"}},"outputId":"54a77c2f-4b87-468e-d2a0-b0ad2e7b2aef"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["SOLEXA-1GA-2_2_FC20EMB:5:251:979:328\t0\t#0\t10145\t25\t36M\t*\t0\t0\tAACCCCTAACCCTAACCCTAACCCTAACCCTAAACT\tarray('B', [71, 71, 71, 71, 39, 66, 54, 71, 71, 39, 51, 70, 71, 66, 42, 32, 62, 46, 45, 71, 32, 32, 36, 36, 33, 57, 36, 30, 39, 30, 34, 33, 34, 30, 35, 32])\t[('NM', 1), ('X1', 1), ('MD', '33A2')]\n","SOLEXA-1GA-2_2_FC20EMB:5:102:214:278\t0\t#0\t10148\t25\t36M\t*\t0\t0\tCCCTAACCCTAACCCTAACCCTAACCCTAACCTAAC\tarray('B', [71, 65, 69, 71, 71, 71, 55, 52, 56, 71, 51, 62, 52, 43, 57, 67, 43, 49, 51, 42, 45, 40, 44, 40, 42, 38, 43, 41, 34, 39, 37, 37, 41, 48, 41, 45])\t[('NM', 0), ('X0', 1), ('MD', '36')]\n","SOLEXA-1GA-2_2_FC20EMB:5:195:284:685\t16\t#0\t10149\t25\t36M\t*\t0\t0\tCCAAACACTAACCCTAACCCTAACCCTAACCTAACC\tarray('B', [29, 27, 29, 33, 31, 29, 30, 29, 30, 35, 29, 29, 30, 33, 30, 35, 29, 35, 33, 34, 30, 36, 31, 33, 35, 39, 32, 42, 34, 36, 42, 36, 49, 43, 46, 46])\t[('NM', 1), ('X1', 1), ('MD', '29A3A2')]\n","SOLEXA-1GA-2_2_FC20EMB:5:35:583:827\t0\t#0\t10150\t25\t36M\t*\t0\t0\tCTAACCCTAAACCTAACCCTAACCCTAACCTAACCA\tarray('B', [71, 71, 54, 62, 55, 60, 44, 55, 45, 46, 39, 48, 48, 54, 44, 40, 43, 39, 38, 40, 37, 44, 41, 38, 41, 43, 34, 37, 38, 38, 41, 32, 42, 40, 36, 39])\t[('NM', 1), ('X1', 1), ('MD', '10A24A0')]\n","SOLEXA-1GA-2_2_FC20EMB:5:248:130:724\t0\t#0\t10152\t25\t36M\t*\t0\t0\tAACCCTAACCCTAACCCCAACCCTAACCTAACCCTA\tarray('B', [71, 66, 71, 47, 71, 66, 62, 62, 62, 66, 54, 50, 58, 53, 49, 46, 65, 49, 55, 35, 51, 46, 52, 50, 41, 43, 55, 46, 32, 31, 43, 38, 37, 44, 34, 31])\t[('NM', 1), ('X1', 1), ('MD', '17C18')]\n"]}]},{"cell_type":"markdown","source":["### SAM fields\n","\n","As you can see from this last example, each SAM record is on a separate line, and it consists of several tab-delimited fields.  The [SAM specification] is the authoritative source for information on what all the fields mean exactly, but here's a brief summary of each of the fields (in order):\n","\n","1. `qname` is the name of the read\n","2. `flags` is a bit field encoding some yes/no pieces of information about whether and how the read aligned\n","3. `rname` is the name of the reference sequence that the read aligned to (if applicable).  E.g., might be \"chr17\" meaning \"chromosome 17\"\n","4. `pos` is the 1-based offset into the reference sequence where the read aligned.\n","5. `mapq` for an aligned read, this is a confidence value; high when we're very confident we've found the correct alignment, low when we're not confident\n","6. `cigar` indicates where any gaps occur in the alignment\n","7. `rnext` only relevant for paired-end reads; name of the reference sequence where other end aligned\n","8. `pnext` only relevant for paired-end reads; 1-based offset into the reference sequence where other end aligned\n","9. `tlen` only relevant for paired-end reads; fragment length inferred from alignment\n","10. `seq` read sequence. If read aligned to reference genome's reverse-complement, this is the reverse complement of the read sequence.\n","11. `qual` quality sequence. If read aligned to reference genome's reverse-complement, this is the reverse of the quality sequence.\n","12. `extras` tab-separated \"extra\" fields, usually optional and aligner-specific but often very important!\n","\n","[SAM specification]: http://samtools.sourceforge.net/SAMv1.pdf\n"],"metadata":{"id":"DWBoTtW59yY3"}},{"cell_type":"code","source":[],"metadata":{"id":"qpuWTFBamM6-"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"H8-_iOurmM3i"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"wtQXhcWumM07"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"2PEGFEN1mMyF"},"execution_count":null,"outputs":[]}]}